### ISSUE TITLE
Handle TAAR failures more gracefully

### ISSUE THREAD

**Issue by diox** [mozilla/addons-server#16182]
Now that TAAR is hosted on GCP, we are depending on two different cloud providers at the same time. We do have a timeout (currently set to 1 second for both socket&response body) for requests to TAAR, and it's good for temporary issues for a single requests but that's not enough when there is a wider GCP / TAAR outage - if enough workers are waiting on TAAR at the same time, we'll be unable to serve other requests.

The bottom line is we should not keep attempting connections to TAAR when it's down/slow. I see two potential ways of handling this:

1) Use a connection pool of a fixed size. When the pool doesn't have an available connection (it's busy waiting for responses), return an error immediately to the view, which in turn returns an error to the client or a fallback.
2) Use some kind of mechanism to mark a given host as dead for x seconds after y failures in a row. When a request is made to an host marked as dead (the x seconds are not up yet), return an error immediately to the view, which in turn returns an error to the client or a fallback.

Both mechanisms could be per thread of maybe even shared across all threads if that works. Note that we don't have to care about cookies at all - it's just pure stateless HTTP requests. Also, every request to TAAR already goes through a single function call, `call_recommendation_server()` (host may differ between TAAR and TAAR-lite however).

Alternatively, we could also do:

3) Do nothing automatically, just add a waffle switch to manually deactivate TAAR-related features. There is one already for discovery endpoint, but the addon recommendation endpoint doesn't have any.

**Comment by bqbn** [comment]
> 3. Do nothing automatically, just add a waffle switch to manually deactivate TAAR-related features. There is one already for discovery endpoint, but the addon recommendation endpoint doesn't have any.

I like this one. If an upstream goes down for an extended period of time, we probably should know. And by which time, human intervention is probably readily available.

No particular preference between 1 and 2 from an ops' perspective, if they are to be implemented.

**Comment by bqbn** [comment]
Last Thursday when we did the push, there was a small uptick in `ConnectTimeout` errors from addons-server to the taar service.

Maybe it's worth keeping this open.

**Comment by diox** [comment]
> It's good for temporary issues for a single requests but that's not enough when there is a wider GCP / TAAR outage

Note that most of our issues with TAAR have been temporary in nature, just a few timeouts per day, not enough to trigger an automated handling of some kind.

For what we were worried about in this issue in particular, a waffle switch would likely work best.

**Comment by diox** [comment]
- Timeout exceptions will no longer go to Sentry (rest of exceptions still caught as normal)
- A ne waffle switch, `enable-taar`, as been added (with a value of `on`). That controls whether or not TAAR/TAAR-lite is enabled or not - when disabled, we show the fallback content.

**Comment by ioanarusiczki** [comment]
@diox 

On AMO dev I tested with the switch enabled and then disabled -> I get the same results. To detail a bit: 

- the[ recommendation](https://addons-server.readthedocs.io/en/latest/topics/api/addons.html#recommendations) endpoint returns 2 add-ons with `outcome:recommended_fallback` , `fallback_reason:no_results` 
- [discovery](https://addons-server.readthedocs.io/en/latest/topics/api/discovery.html) endpoint returns every time the [admin discovery items 
](https://addons-internal.dev.mozaws.net/en-US/admin/models/discovery/discoveryitem/) which I assume is the "static curated content" -> all results return ` is_recommendation: false`

Right now on AMO prod: 
- recommendations on add-on detail page change ,confirmed by `outcome:recommended` in the response
- discovery items are changing, I see is_recommendation:true in responses (the addon is returned by TAAR). There are 5 extensions and 3 static themes which don't change (that's expected according to documentation) 

I am not sure how to check this issue on -dev. If I remember well, there's never been TAAR content on the testing servers.

**Comment by diox** [comment]
TAAR does know how to return results on dev, but the problem is that the add-on it returns need to exist and the right client id/guid needs to be used. The rules are:
- be public and recommended, and be more than 4, for the add-on recommendation endpoint (https://addons-dev.allizom.org/api/v5/discovery/?telemetry-client-id=11111111-1111-1111-1111-111111111111)
- be public and have a discovery item, for the discovery endpoint (https://addons-dev.allizom.org/api/v5/addons/recommendations/?guid=uBlock0@raymondhill.net&recommended=true)

To meet the conditions on dev:
- All the following add-ons should exist, be public and recommended `'jid1-MnnxcxisBPnSXQ@jetpack', 'jid1-BoFifL9Vbdl2zQ@jetpack', 'firefox@betterttv.net', 'sponsorBlocker@ajay.app'`
- A few (results are random, so we need enough to guarantee enough working ones will be present) of the following add-ons should exist, be public and have a discovery item: `'addon@darkreader.org', 'adguardadblocker@adguard.com', 'clearcache@michel.de.almeida', 'copyplaintext@eros.man', 'default-bookmark-folder@gustiaux.com', 'enhancerforyoutube@maximerf.addons.mozilla.org', 'extension@one-tab.com', 'extension@tabliss.io', 'forecastfox@s3_fix_version', 'foxyproxy@eric.h.jung', 'foxytab@eros.man', 'gmailnoads@mywebber.com', 'jid0-bnmfwWw2w2w4e4edvcdDbnMhdVg@jetpack', 'jid1-BoFifL9Vbdl2zQ@jetpack', 'jid1-Kt2kYYgi32zPuw@jetpack', 'jid1-ZAdIEUB7XOzOJw@jetpack', 'juraj.masiar@gmail.com_ScrollAnywhere', 'printedit-we@DW-dev', 'tabby@whatsyouridea.com', 'uBlock0@raymondhill.net', '{446900e4-71c2-419f-a6a7-df9c091e268b}', '{7b1bf0b6-a1b9-42b0-b75d-252036438bdc}', '{a9c2ad37-e940-4892-8dce-cd73c6cbbc0c}', '{b9db16a4-6edc-47ec-a1f4-b86292ed211d}', '{c2c003ee-bd69-42a2-b0e9-6f34222cb046}', '{d07ccf11-c0cd-4938-a265-2a4d6ad01189}'`

Looks like most of these exist but are necessarily public/recommended/with a discovery item.

**Comment by diox** [comment]
Sorry, the test client id to use for that second endpoint is actually `1111111111111111111111111111111111111111111111111111111111111111` (needs to be a fake hash of 64 characters, all of them being the same digit)

**Comment by ioanarusiczki** [comment]
@diox 

Ok, thanks, so I made several requests to https://addons-dev.allizom.org/api/v5/discovery/?telemetry-client-id=1111111111111111111111111111111111111111111111111111111111111111

I started getting results with is_recommendation true , returning addons which are part of the [discovery list](https://addons-internal.dev.mozaws.net/en-US/admin/models/discovery/discoveryitem/) but also other addons, I'll give 3 examples:
https://addons-dev.allizom.org/en-US/firefox/addon/swift-selection-search/
https://addons-dev.allizom.org/en-US/firefox/addon/s3download-statusbar/
https://addons-dev.allizom.org/en-US/firefox/addon/privacy-recommended/

I hope that's alright. 

After disabling switch I get "is_recommendation": false and the [discovery add-ons](https://addons-internal.dev.mozaws.net/en-US/admin/models/discovery/discoveryitem/) with positions set are returned.

And I tested again the recommendations -> same results as I described above. I assume it's ok, I still don't understand why the forth add-on won't be `sponsorBlocker@ajay.app`

**Comment by diox** [comment]
> And I tested again the recommendations -> same results as I described above. I assume it's ok, I still don't understand why the forth add-on won't be `sponsorBlocker@ajay.app`

Sorry, that's my fault: actually the add-ons returned by this endpoint don't need to have a recommended promotion at all, they just need to exist and be public. So here TAAR is returning that fourth add-on and since it exists, the endpoint is returning it as expected.

**Comment by ioanarusiczki** [comment]
@diox Then the same should happen for the discovery endpoint, reason why I got public add-ons which are not part of the list. 

I think I can mark this verified.

**Comment by KevinMind** [comment]
Old Jira Ticket: https://mozilla-hub.atlassian.net/browse/ADDSRV-50
